Data driven tests

mattphillips opened the pull request:
## Summary

Add Data Driven Testing from [`jest-each`](https://github.com/mattphillips/jest-each) as discussed in #6082 

## Test plan

I've added tests using the new bindings. I've only added one test for the `skip` and `only` bindings as otherwise they will result in the runner appearing to ignore multiple tests.

`jest-each` is also 100% unit tested under the hood too.

| New API              | `jest-each` equivalent |
| ------------------ |----------------------|
| `test.each`          | `each.test`            |
| `xtest.each`         | `each.xtest`           |
| `it.each`            | `each.it`              |
| `describe.each`      | `each.describe`        |
| `fit.each`           | `each.fit`             |
| `xdescribe.each`     | `each.xdescribe`       |
| `fdescribe.each`     | `each.fdescribe`       |
| `xit.each`           | `each.xit`             |
| `test.only.each`     | `each.test.only`       |
| `test.skip.each`     | `each.skip.only`       |
| `it.only.each`       | `each.it.only`         |
| `it.skip.each`       | `each.it.skip`         |
| `describe.only.each` | `each.describe.only`   |
| `describe.skip.each` | `each.describe.skip`   |

### TODO

- [ ] Add docs to website

/cc @SimenB

SimenB commented:
This looks awesome, thanks so much for doing this!

@cpojer thoughts on the dependency over importing the source code?

If we do keep it as a dep, I think we should move it to `jest-community`.

>I've added tests using the new bindings. I've only added one test for the `skip` and `only` bindings as otherwise they will result in the runner appearing to ignore multiple tests.

I think we should add some integrations tests, both to assert on the output, but also to see that `skip`, `only` and failures looks correct.

cpojer commented on packages/jest-jasmine2/src/index.js:
```diff
@@ -70,6 +71,40 @@ async function jasmine2(
+        [parent][child](title, test);
+  };
+
+  [
```
There is only a finite set of these. It's more to type out, but can we be less smart and just manually inline this, like:

```
environment.global.test.each = …
environment.global.xtest.each = …
```

etc.?

If you look at line 67
https://github.com/facebook/jest/pull/6102/files#diff-e4399d1682ddb42f8a43f1a917d42d87R67

You'll also see that a lot of these are just aliases for each other, so you don't need to assign `.each` to all of them, just about half of them! :)

mattphillips replied:
Good shout, I've just updated :)

cpojer commented:
If this is gonna be a core part of Jest, it's got to be inside of this repo, assuming @mattphillips is happy to contribute `jest-each` to this repo and update it to the standards of this repo.

SimenB commented on packages/jest-jasmine2/package.json:
```diff
@@ -14,6 +14,7 @@
     "graceful-fs": "^4.1.11",
     "is-generator-fn": "^1.0.0",
     "jest-diff": "^22.4.0",
+    "jest-each": "^0.5.0",
```
To easier roll this in long-lived jest releases, it would be awesome to release 1.0.0 before merging this 🙂

mattphillips commented:
My bad I misunderstood the comments in the issue I thought the idea was to import `jest-each` as a dep. I'm happy to merge it into core, I suspect by doing that `jest-each`'s code can be simplified, when I made it I didn't want to pollute Jest's prototype, so `jest-each` acts as a proxy to existing APIs.

If it is part of core it will be fine to add these additional properties without the need to proxy too :smile:.

What is the process for moving `jest-each` into core?

SimenB commented:
If you are fine with not supporting jest@<23 moving forward, I'd say move it into `jest-jasmine2` and extend as needed by adding the properties directly.
If you want to keep supporting jest<23, then `jest-each` as is (pretty much, it needs to be typed using flow etc) should be made a separate package in this repo.

My only worry is that it will make `jest-circus` drift further from `jest-jasmine2`, but it shouldn't be too much of a bother fixing that later.

mattphillips commented:
> One advantage of just moving it into jest-jasmine2 without making it a separate package is that you can keep maintaining jest-each as a package for older versions of jest. That might actually be the best solution, come to think of it 🙂

I'm open to either solution :smile:

I think it could lead to a situation where people raise issues in the wrong place potentially? `Why is .each not working?` in core and then we have to send them over to `jest-each`. Although that could be a moot point as we'd still just tell them to download `jest-each` even though its hosted in core 🤷‍♂️ 

Hmmm actually the biggest issue I see is keeping them in sync, bug fixes would generally apply to both the `jest-jasmine2` implementation and `jest-each`.

I'm now leaning towards bring the existing lib into core 😆

cpojer commented:
I think directly embedding into jest-jasmine2 makes sense. We'll deal with circus separately.

SimenB commented on packages/jest-jasmine2/package.json:
```diff
@@ -19,7 +19,8 @@
     "pretty-format": "^22.4.0",
-    "source-map-support": "^0.5.0"
+    "source-map-support": "^0.5.0",
+    "sprintf-js": "^1.1.1"
```
do we need this when we have https://nodejs.org/api/util.html#util_util_format_format_args?

mattphillips replied:
Nice I didn't know about `util.format` just updated :)

SimenB commented on packages/jest-jasmine2/src/__tests__/each.test.js:
```diff
@@ -0,0 +1,266 @@
+
+          expect(() =>
+            globalMock.mock.calls[0][1](),
+          ).toThrowErrorMatchingSnapshot();
```
instead of doing this, I wonder if it would be better to create actual integration tests (in https://github.com/facebook/jest/tree/master/integration-tests) and assert on the output there?

While this works, it seems sorta weird to invoke something passed to a mock function, and making asserts on how that behaves.

Might just be me, though!

mattphillips replied:
I agree it would be nice to have an integration test too, I used TDD when building this thats why it seems strange but it allowed me to just write unit tests for the behaviour 😃 

I'll add some tests in /integration-tests too (I didn't know about that directory)

SimenB commented on packages/jest-jasmine2/src/each.js:
```diff
@@ -0,0 +1,84 @@
+/**
```
This should have the full license header, methinks.

```
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
```

SimenB replied:
not sure about that one, though...

mattphillips replied:
Yeah I wasn't sure either - I can add it to be on the safe side

mattphillips replied:
Done in 30ebfb5

SimenB commented on integration-tests/each/__tests__/each-exception.test.js:
```diff
@@ -0,0 +1,17 @@
+/**
```
you'll also need to execute them (and assert on the output). See e.g. https://github.com/facebook/jest/blob/master/integration-tests/__tests__/failures.test.js

mattphillips replied:
🤦‍♂️ I was wondering how they were ran 🤣

mattphillips replied:
Added in 6607413

mattphillips commented:
@SimenB @cpojer I think this is at a good point now, if you guys could take another look :smile:

Only outstanding thing is to write up some docs for the website.

SimenB commented:
Seems like one of the new tests are failing CI

SimenB commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ### Features
 
+* `[jest-jasmine2]` Add data driven testing with `jest-each`
```
maybe tweak this a bit.

SimenB commented on integration-tests/__tests__/__snapshots__/each.test.js.snap:
```diff
@@ -0,0 +1,212 @@
+    Not enough arguments supplied for given headings: left | right
+    Received: true,true,true
+      
+      at packages/jest-jasmine2/build/each.js:75:17
```
I wonder if this message could be better (it should also have a stack pointing back to the user code.)

SimenB replied:
maybe using `pretty-format` for the `Received: true,true,true` part, and explicitly saying "expected length 2, got 3" somehow

mattphillips replied:
I'm open to suggestions as to what it should say.

I'll try to add the stack too

SimenB replied:
I added some suggestions in a second comment, but the error could contain the line that is wrong as well?

SimenB commented on integration-tests/__tests__/__snapshots__/each.test.js.snap:
```diff
@@ -0,0 +1,212 @@
+  ✓ passes one row expected true == true
+  ✓ passes one row expected true == true
+  ✓ passes one row expected true == true
+  ○ skipped 4 tests
```
is `4` correct? maybe not possible to make better

mattphillips replied:
Yeah there is 8 tests, 4 with `.only` so 4 are skipped

Same for the `.skip` test but just inversed

SimenB commented on integration-tests/__tests__/each.test.js:
```diff
@@ -0,0 +1,74 @@
+  const rest = extractSummary(result.stderr)
+    .rest.split('\n')
+    .filter(line => line.indexOf('packages/expect/build/index.js') === -1)
+    .join('\n');
```
Do you need the `split` etc? It's a special case of `failures`

mattphillips replied:
Nope removed

SimenB commented on integration-tests/__tests__/each.test.js:
```diff
@@ -0,0 +1,74 @@
+const {extractSummary} = require('../Utils');
+const dir = path.resolve(__dirname, '../each');
+
+SkipOnWindows.suite();
```
it would be great if this was not needed. if it is, no biggie

mattphillips replied:
I can't see why it wouldn't work on windows so I'll remove it and see what happens :)

SimenB commented on integration-tests/__tests__/each.test.js:
```diff
@@ -0,0 +1,74 @@
+  const rest = extractSummary(result.stderr)
+    .rest.split('\n')
+    .filter(line => line.indexOf('packages/expect/build/index.js') === -1)
+    .map(line => line.trim())
```
it'd be great to not trim, so the output in the snapshot matches stdout

mattphillips replied:
Cool removed it

SimenB commented:
This is really shaping up! I think it mainly just misses docs and perhaps some improvements for the error messages to help the user understand what failed. Tweaks like that can come later though, when people play around with it 🙂

mattphillips commented:
Hey @SimenB CI is still failing for a snapshot that I've updated and pushed 😕 I'm not sure how to fix it? The error in the snapshot is the same too 😆 

<img width="906" alt="screen shot 2018-05-01 at 15 26 06" src="https://user-images.githubusercontent.com/5610087/39476531-29e7f70c-4d54-11e8-99db-d0866ee897f6.png">

SimenB commented:
We need to skip on windows, as the checkmarks and x-es are different.

But it seems like it also highlights your issue - is there some trailing spaces missing somewhere?

![image](https://user-images.githubusercontent.com/1404810/39481295-3c0d52ee-4d6b-11e8-935f-1511922997a3.png)

Although since node 6 passes that seems weird.

mattphillips commented:
@SimenB I've just added skipping on windows and updated the snapshot again hopefully it'll pass this time (3rd time lucky!)

I've also added some docs to the website if you wouldn't mind taking a look over them :smile:

Thanks a lot for all of your help 🥇

mattphillips commented:
@SimenB @cpojer I think this is good to go if you guys could review again - thanks!

SimenB commented:
Could you rebase? I think that might fix CI 🙂

SimenB commented on docs/GlobalAPI.md:
```diff
@@ -256,6 +256,81 @@ describe('binaryStringToNumber', () => {
+    });
+
+    test('does not mutate first arg', () => {
+      a + b;
```
what does this line do?

mattphillips replied:
Was a bad example test, I've just updated to something simpler :)

SimenB commented on docs/GlobalAPI.md:
```diff
@@ -278,6 +353,52 @@ describe('my other beverage', () => {
+### `describe.only.each(table)(name, fn)`
+
+Also under the aliases: `fdescribe.each(table)(name, fn)` and
+`` fdescribe.each`table`(name, fn) ``
```
extra space?

mattphillips replied:
Prettier puts them in

mattphillips commented:
@SimenB it’s all rebased and passing now 🎉 thanks again for your help and comments 🙂

@cpojer fancy taking a look over this?

cpojer commented:
Awesome work. Thanks so much for contributing this project to Jest, this is really cool.

SimenB commented:
It'd be interesting to gauge interest in prettier for fixing the template strings

mattphillips commented:
@SimenB I agree we should look into prettier support. What does styled-components or other libraries that use tagged templates do for code formatting? 

Worst case scenario maybe we write an eslint rule with a fix option

SimenB commented:
styled-components is built-in to prettier, along with graphql (the 2 I know using tagged templates)
